src/Disks/ObjectStorages/MetadataStorageFromPlainObjectStorageOperations.cpp

```diff
@@ -183,18 +183,24 @@ void MetadataStorageFromPlainObjectStorageMoveDirectoryOperation::execute(std::u
     constexpr bool validate_content = false;
 #endif
 
-    auto write_buf = createWriteBuf(path_from, path_to, validate_content);
-    writeString(path_to.string(), *write_buf);
-
-    fiu_do_on(FailPoints::plain_object_storage_write_fail_on_directory_move,
+    std::unordered_set<std::string> subdirs = {""};
+    path_map.iterateSubdirectories(path_from.parent_path().string() + "/", [&](const auto & elem){ subdirs.emplace(elem); });
+    for (const auto & subdir : subdirs)
     {
-        throw Exception(ErrorCodes::FAULT_INJECTED, "Injecting fault when moving from '{}' to '{}'", path_from, path_to);
-    });
+        auto write_buf = createWriteBuf(path_from / subdir, path_to / subdir, validate_content);
+        writeString((path_to / subdir).string(), *write_buf);
 
-    write_buf->finalize();
+        fiu_do_on(FailPoints::plain_object_storage_write_fail_on_directory_move,
+        {
+            throw Exception(ErrorCodes::FAULT_INJECTED, "Injecting fault when moving from '{}' to '{}'", path_from, path_to);
+        });
+
+        write_buf->finalize();
+
+        /// parent_path() removes the trailing '/'.
+        path_map.moveDirectory((path_from / subdir).parent_path(), (path_to / subdir).parent_path());
+    }
```

Comment:
`MetadataStorageFromPlainObjectStorageMoveDirectoryOperation::undo` should also be changed to revert to the previous state if rollback is triggered, but this is not very important; I can follow up separately.

Reply:
A `LOG_TEST` message that a subdirectory was moved might be useful.